File: src/main/java/seremis/geninfusion/item/GIItem.java
package seremis.geninfusion.item;

import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.renderer.texture.IIconRegister;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.IIcon;
import seremis.geninfusion.GeneticInfusion;
import seremis.geninfusion.lib.DefaultProps;

import java.util.List;

public class GIItem extends Item {

    private IIcon iconBuffer[];
    private int metadata = 0;

    public GIItem() {
        super();
        setMaxStackSize(64);
        setCreativeTab(GeneticInfusion.CreativeTab);
    }

    @Override
    public GIItem setUnlocalizedName(String par1Str) {
        return (GIItem) super.setUnlocalizedName(par1Str);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void registerIcons(IIconRegister iconRegister) {
        if(this.metadata == 0) {
            itemIcon = iconRegister.registerIcon(DefaultProps.ID + ":" + this.getUnlocalizedName().substring(5));
        } else {
            iconBuffer = new IIcon[metadata];
            for(int x = 0; x < iconBuffer.length; x++) {
                iconBuffer[x] = iconRegister.registerIcon(DefaultProps.ID + ":" + this.getUnlocalizedName().substring(5) + (x + 1));
            }
        }
    }

    @Override
    @SideOnly(Side.CLIENT)
    public IIcon getIconFromDamage(int metadata) {
        if(this.metadata != 0) {
            itemIcon = iconBuffer[metadata];
        }
        return this.itemIcon;
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void getSubItems(Item item, CreativeTabs creativetab, List list) {
        if(metadata > 0) {
            for(int i = 0; i < getNumbersofMetadata(); ++i) {
                list.add(new ItemStack(item, 1, i));
            }
        } else {
            list.add(new ItemStack(item, 1, 0));
        }
    }

    public int getNumbersofMetadata() {
        return this.metadata;
    }

    public void setNumbersofMetadata(int metadata) {
        this.metadata = metadata;
    }

}

File: src/main/java/seremis/geninfusion/item/GIItem.scala
package seremis.geninfusion.item

import java.util

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.client.renderer.texture.IIconRegister
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.{Item, ItemStack}
import net.minecraft.util.IIcon
import seremis.geninfusion.GeneticInfusion
import seremis.geninfusion.lib.DefaultProps

class GIItem extends Item() {

    setMaxStackSize(64)
    setCreativeTab(GeneticInfusion.creativeTab)

    var iconBuffer: Array[IIcon] = null
    var metadata: Int = 0

    override def setUnlocalizedName(par1Str: String): GIItem = {
        super.setUnlocalizedName(par1Str).asInstanceOf[GIItem]
    }

    @SideOnly(Side.CLIENT)
    override def registerIcons(iconRegister: IIconRegister) {
        if (this.metadata == 0) {
            itemIcon = iconRegister.registerIcon(DefaultProps.ID + ":" + this.getUnlocalizedName.substring(5))
        } else {
            iconBuffer = Array.ofDim[IIcon](metadata)
            for (x <- 0 until iconBuffer.length) {
                iconBuffer(x) = iconRegister.registerIcon(DefaultProps.ID + ":" + this.getUnlocalizedName.substring(5) +
                    (x + 1))
            }
        }
    }

    @SideOnly(Side.CLIENT)
    override def getIconFromDamage(metadata: Int): IIcon = {
        if (this.metadata != 0) {
            itemIcon = iconBuffer(metadata)
        }
        this.itemIcon
    }

    @SideOnly(Side.CLIENT)
    override def getSubItems(item: Item, creativetab: CreativeTabs, list: util.List[_]) {
        if (metadata > 0) {
            for (i <- 0 until getNumbersofMetadata) {
                list.asInstanceOf[util.List[ItemStack]].add(new ItemStack(item, 1, i))
            }
        } else {
            list.asInstanceOf[util.List[ItemStack]].add(new ItemStack(item, 1, 0))
        }
    }

    def getNumbersofMetadata: Int = this.metadata

    def setNumbersofMetadata(metadata: Int) {
        this.metadata = metadata
    }
}